### DISCUSSION TITLE
Flow & type-safety

### DISCUSSION THREAD

**Discussion by arcanis** [mobxjs/mobx#2959] [Ideas]
There's a known issue with TypeScript and Generator-based APIs in that the `yield` expressions are typed `any` (https://github.com/mobxjs/mobx/discussions/2556). I just wanted to let you know that I've submitted a proposal (and a candidate implementation) for a fix, which you can find tin the following thread: https://github.com/microsoft/TypeScript/issues/43632

Under this proposal, you'd be able to annotate the yield expression results using the special `YieldType` marker, which would instruct TypeScript that, when passed through a `yield` expression, the returned value is the generic parameter:

```ts
declare function fetchGithubProjectsSomehow(): Promise<Project[]> & YieldType<Project[]>;

class Store {
    *fetchProjects() {
        this.githubProjects = yield fetchGithubProjectsSomehow();
    }
}
```

At the moment the [current blocker](https://github.com/microsoft/TypeScript/issues/43632#issuecomment-849756087) requires to show the TypeScript team that this feature would have a positive practical impact, so if that's something you're interested to ever see fixed I'd appreciate any feedback you could share! 🙂

**Comment by iChenLei** [comment]
Cool, thanks for your work.
